src/crimson/os/seastore/onode_manager/staged-fltree/node.cc

```diff
@@ -683,22 +683,29 @@ eagain_future<Ref<Node>> Node::load(
              c.t, addr, expect_is_level_tail);
       ceph_abort("fatal error");
     })
-  ).safe_then([FNAME, c, expect_is_level_tail](auto extent) {
+  ).safe_then([FNAME, c, expect_is_level_tail](auto extent)
+    -> eagain_future<Ref<Node>> {
+    //XXX: this should be removed once interruptible future is involved
+    if (!extent->is_valid()) {
+      return crimson::ct_error::eagain::make();
+    }
```

Comment:
Seems to me this is a duplicated check. So can this be removed?

Reply:
I don't think we need check if extent is valid in node.cc, extent can be invalid at any time during the transaction. we do duplicate_for_write to handle the confliction.

Reply:
Actually, I think this is like dealing pg interval change at the pg level. As even successive continuations can be executed asynchrously, which means states can be different in successive continuations, every continuation needs to take care of critical state change, which is why we need something like interruptible future.

Being specific to this case, even if the extent is valid when we do `duplicate_for_write()`, it can have been invalidated by other transactions when the execution reaches here, so I think we also need to take care of it. Am I missing anything?

Reply:
> it can have been invalidated by other transactions when the execution reaches here

`(Seastore)NodeExtentManager::read_extent()` already has the same check **after** the async read operation, so seems to me this is duplicated as it there is no async operations between the previous check. Is it correct?

Reply:
Um, I looks not to me. When async continuations in (`Seastore)NodeExtentManager::read_extent()` finishes execution, it just resolve the corresponding future, which further leads to reactors put the successvie continuation onto the execution queue, but doesn't invoke that successive continuation directly. Am I right?

Reply:
> but doesn't invoke that successive continuation directly.

IIUC, for the available futures, it does invoke the linked lambda directly, see: https://github.com/scylladb/seastar/blob/2e3c6da3f8c8b69cb790b056661717f1d7346a61/include/seastar/core/future.hh#L1597-L1603

And in this case, the lambda implementing the first check returns the available future, see:
https://github.com/ceph/ceph/blob/a52511c1bac3bcc2e702b9ca15e38afa027130fd/src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h#L114-L121

Reply:
I think all continuations following an async continuation will all be executed asynchronously.  Here is the async continuation: https://github.com/ceph/ceph/blob/a52511c1bac3bcc2e702b9ca15e38afa027130fd/src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h#L108. So later continuations will be executed asynchronously.

Reply:
> > but doesn't invoke that successive continuation directly.
> 
> IIUC, for the available futures, it does invoke the linked lambda directly, see: https://github.com/scylladb/seastar/blob/2e3c6da3f8c8b69cb790b056661717f1d7346a61/include/seastar/core/future.hh#L1597-L1603

I think continuations can only be executed synchronously if the future is resolved at the moment that` future::then()` is invoked.

Reply:
It is fine for me to keep the second check if insisted, but I think this is a good chance to understand seastar assumptions about atomic operations.

To simplify the logic we are discussing:
```
return seastar::now().then([t, addr] {
  // continuation A
  return TransactionManager::read_extent<SeastoreNodeExtent>(t, addr);
)).safe_then(auto&& e) {
  // continuation B
  if (!e->is_valid()) {
    return crimson::ct_error::eagain::make();
  } else {
    return NodeExtentRef(e);
  }
}).safe_then(auto&& e) {
  // continuation C
  if (!e->is_valid()) {
    return crimson::ct_error::eagain::make();
  }
  {... other logic ...}
});
```

I think the check `!e->is_valid()` in C is not necessary because B has the same check and returns an immediate future, which means B is "atomically" linked with C whenever `e` is valid. IIUC it should be impossible for any other continuations to be able to change the status of `e` between B and C.

Reply:
> I think the check `!e->is_valid()` in C is not necessary because B has the same check and returns an immediate future, which means B is "atomically" linked with C whenever `e` is valid. IIUC it should be impossible for any other continuations to be able to change the status of `e` between B and C.

Actually, it looks to me that B is not atomically linked with C, B returning an available future only cause C to be put onto seastar reactor's execution queue.

Reply:
Hmm, @tchaikov @athanatos what do you think about https://github.com/ceph/ceph/pull/41867#discussion_r653302802?

Reply:
Don't rely on the two safe_then continuations occurring atomically -- check both times.

Reply:
when we check if child node is valid the parent node maybe invalid also by other transaction. I don't think it is a good mechanism to check extent invalid in each continuation.  Is it a problem when we use invalid(retired) node to construct record? @athanatos

Reply:
Right, you'd have to check every extent referenced in a continuation after it was obtained.  However, this will be handled automatically and transparently by the interruptible_future change I'm doing seastore-wide.

Reply:
yep, that will be reasonable, automatically and transparently check for extent's user, then the user needn't check it by itself.

Reply:
yeah, per our discussion offline, we should not rely on the atomicity. it does *not* exist at all, if the continuation does not return an immediately-available future, and this pushes all continuations in the chain after it to task queue, so they have to be scheduled and executed by reactor in an asynchronous manner. so we cannot assume that they are performed atomically. but if the first one in the chain is completed in a non-blocking call, `then` would try to evaluate the following call to see if the future returned by it is valid or not. but again, we cannot explore this as a way to structure our code. because, if we can ensure the first the valid state of the first future, why don't we just merge the first continuation right into the second one?

Reply:
I'm not aware of any place in our code that assumes atomicity across immediately-availble futures. Thanks for the feedbacks!